### PR TITLE
Add collection method to decorate collections of objects

### PR DESCRIPTION
**Pull request by wout**
This PR adds the possibility to use collections as well:

```crystal
UserDecorator.collection([User.new]).first.full_name
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -69,6 +72,15 @@ puts decorated_time.date
 # => "November 2, 2020"
 ```
 
+Or even a collection of time objects with the `collection` method:
+
+```crystal
+decorated_times = TimeDecorator.collection([Time.utc])
```
Comment (stephendolan) [inline README.md]: Can you update your example to show what the output will be with multiple `Time` objects?

File: spec/decorator_spec.cr [hunk 2]
```diff
@@ -27,10 +27,14 @@ describe Decorator do
     UserDecorator.new(User.new).website_url.should eq "https://luckycasts.com"
   end
 
-  it "successfully decorates objects" do
+  it "successfully decorates an object" do
     UserDecorator.new(User.new).full_name.should eq "Lucky Casts"
   end
 
+  it "successfully decorates a collection of objects" do
+    UserDecorator.collection([User.new]).first.full_name.should eq "Lucky Casts"
```
Comment (stephendolan) [inline spec/decorator_spec.cr]: Can you add another test to show that a collection with multiple items also works as expected?

**Comment by stephendolan** [comment]
Thanks for this, @wout!

Could you update your PR summary with some examples of how you're planning to use this? I'm having trouble envisioning when you would leverage this outside of an `each` loop when you'd already have access to the individual object you're decorating.

**Comment by wout** [comment]
For example, when you have an app rendering ERC views and you don't want any logic in them, then it's convenient to have a pre-decorated collection.

**Comment by stephendolan** [comment]
Cool. Yeah, if you could drop a few code blocks in the PR description for reference to show how you're going to use this, that'd be really helpful!

I often reference or go back to PRs when I'm wondering about pieces of functionality, and I think that context is important for this one.

**Comment by wout** [comment]
Absolutely, I'll drop some here tomorrow.

**Comment by wout** [comment]
Sorry, it has taken a bit longer to get back here.

So, I have a fairly generic component in a Lucky app for rendering tables with arbitrary items. Here's a simplified version:

```cr
class Shared::Admin::ItemTable(T) < BaseComponent
  needs items : Array(T)

  def render
    items.any? ? render_table : render_no_items
  end

  private def render_table
    {% begin %}
      {% columns = T.annotation(BaseDecorator::Table)[:columns] %}

      table do
        items.each do |item|
          tr do
            {% for column in columns %}
              td item.{{column.id}}.to_s
            {% end %}
          end
        end
      end
    {% end %}
  end
  # ...
end
```

The type var `T` here is a decorator, and the given decorator is annotated with the necessary attributes for the compiler:

```cr
@[BaseDecorator::Table(columns: %i[email name role])]
struct UserDecorator < BaseDecorator
  decorates user : User

  def name
    "#{first_name} #{last_name}"
  end

  def role
    user.role.name
  end
end
```

Finally, the component is fed with an array of decorators using the `UserDecorator.collection` class method:

```cr
class AdminUsers::IndexPage < AdminLayout
  needs users : UserQuery

  def content
    mount Shared::Admin::ItemTable(UserDecorator),
      items: UserDecorator.collection(users.to_a)
  end
end
```

This is one example, but I can think of more scenarios where the `collection` method may come in handy. You could argue to use the native `map` method everywhere, but that would result in a lot of repetition.

**Comment by stephendolan** [comment]
Thanks, @wout !

If you can address the few comments I left in the code, this will be good to merge and cut a release 🥳

**Comment by wout** [comment]
Great, I'd love to. But I don't see any comments. Or am I completely missing something?  :zany_face:

**Review by stephendolan — CHANGES_REQUESTED** [review]
Oops, I didn't realize I had to submit my review for the comments to show up 😄

File: README.md (hunk 1 above)
Comment (wout) [reply README.md]: Of course, that's done!

File: spec/decorator_spec.cr (hunk 2 above)
Comment (wout) [reply spec/decorator_spec.cr]: And done as well.

**Review by stephendolan — APPROVED** [review]

**Pull request merged**